src/commands/amend.rs

```diff
@@ -0,0 +1,128 @@
+    };
+    let head_commit = repo.find_commit_or_fail(head_oid)?;
+
+    let index = repo.get_index()?;
+    if index.has_conflicts() {
+        writeln!(
+            effects.get_output_stream(),
+            "Cannot amend, because there are unresolved merge conflicts."
+        )?;
+        return Ok(1);
+    }
+
+    let staged_index_paths = repo.get_staged_paths()?;
+    let opts = if !staged_index_paths.is_empty() {
+        let dirty_working_tree = repo.has_changed_files(effects, git_run_info)?;
+        if dirty_working_tree {
+            writeln!(
+                effects.get_output_stream(),
+                "Found staged files in the index. Unstaged changes will not be amended."
+            )?;
+        }
+        AmendFastOptions::FromIndex {
+            paths: staged_index_paths.into_iter().collect_vec(),
+        }
+    } else {
+        let status = repo.get_status(git_run_info)?;
+        let entries_to_amend = status
+            .into_iter()
+            .filter(|entry| match entry.working_copy_status {
+                FileStatus::Added
+                | FileStatus::Copied
+                | FileStatus::Deleted
+                | FileStatus::Modified
+                | FileStatus::Renamed => true,
+                FileStatus::Ignored
+                | FileStatus::Unmerged
+                | FileStatus::Unmodified
+                | FileStatus::Untracked => false,
+            })
+            .collect_vec();
+        AmendFastOptions::FromWorkingTree {
+            status_entries: entries_to_amend,
+        }
+    };
+    if opts.is_empty() {
+        writeln!(effects.get_output_stream(), "Nothing to amend.")?;
+        return Ok(0);
+    }
+
+    let amended_tree = repo.amend_fast(&head_commit, &opts)?;
+    let timestamp = now.duration_since(SystemTime::UNIX_EPOCH)?.as_secs_f64();
+
+    let (author, committer) = (head_commit.get_author(), head_commit.get_committer());
+    let (author, committer) = if get_restack_preserve_timestamps(&repo)? {
+        (author, committer)
+    } else {
+        (
+            author.update_timestamp(now)?,
+            committer.update_timestamp(now)?,
+        )
+    };
+
+    let amended_commit_oid = head_commit.amend_commit(
+        Some("HEAD"),
+        Some(&author),
+        Some(&committer),
+        None,
+        Some(&amended_tree),
+    )?;
+    mark_commit_reachable(&repo, amended_commit_oid)
+        .wrap_err("Marking commit as reachable for GC purposes.")?;
+
+    let event_tx_id = event_log_db.make_transaction_id(now, "amend")?;
+    event_log_db.add_events(vec![Event::RewriteEvent {
+        timestamp,
+        event_tx_id,
+        old_commit_oid: head_oid.into(),
+        new_commit_oid: amended_commit_oid.into(),
+    }])?;
+
+    // Reset the index
+    git_run_info.run(effects, Some(event_tx_id), &["reset"])?;
+    git_run_info.run(
+        effects,
+        Some(event_tx_id),
+        &["checkout", &amended_commit_oid.to_string()],
+    )?;
```

Comment:
What does calling `git reset` do here? And what does checking out the commit do? I thought the point of passing `HEAD` above was that it updated the reference for us, so we wouldn't have to check anything out ourselves.

Can we avoid calling `git reset` and `git checkout` here? Both of those require checking the working copy again, which slows down the operation. We can leave it to a future commit if you don't want to do it in this one.

Reply:
It looks like the `git checkout` is no longer necessary -- I think I added that when I was experimenting with not updating `HEAD` as part of the amend.

`git reset` is used to reset the index following the amend. Without this, when you amend something from the working copy, the index retains a "shadow" diff following the amend. The index and working copy contain inverse diffs of each other that get cancelled out by a `reset`. It's rather strange (which likely indicates I'm missing something here...)

To repro this, remove the checkout line and run:

```
$ echo "test" > test.txt
$ git commit -am "test"
$ echo "updated contents" > foo.txt
$ git amend
$ git status
HEAD detached from afdaf9a
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   foo.txt

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   foo.txt

$ git diff
diff --git a/foo.txt b/foo.txt
index 014fd71..1385f26 100644
--- a/foo.txt
+++ b/foo.txt
@@ -1 +1 @@
-test
+updated contents

$ git diff --cached
diff --git a/foo.txt b/foo.txt
index 1385f26..014fd71 100644
--- a/foo.txt
+++ b/foo.txt
@@ -1 +1 @@
-updated contents
+test
```

---


Note that this _doesn't_ happen if you amend a change that is staged to the index -- only when you amend something unstaged from the working copy.

Reply:
Moved this conversation to #201